### ISSUE TITLE
Options to choose different LaTeX rendering services (p.e. with svg output) and offline conversion with MathJax

### ISSUE THREAD

**Issue by alexpacini** [adam-p/markdown-here#144]
Hi,
I would like to suggest some enhancements to the LaTeX section.
- A drop-down menu to choose from multiple services: I suggest to add

```
<img src="https://latex.codecogs.com/svg.latex?{urlmathcode}" alt="{mathcode}">
```

It works pretty well and the output is in _svg_ format.
- I would like to have the possibility to choose from _pure_ TeX delimiters as  
  `$ ... $` and `$$ ... $$` for the centred environment  
  and _pure_ LaTeX delimiters as  
  `\( ... )\` and `\[ ... \]` for the centred environment
- Can you strip out the _white_ space between the opening delimiter and the first LaTeX formula character? (If I leave the space, I don't get the formula rendered)
- Is it possible to use MathJax for converting LaTeX to svg? (or even to another format, for example chosen by a check box).  
  I have seen something interesting on [MathJax's Website](http://docs.mathjax.org/en/latest/configuration.html#using-a-configuration-file) especially with the configuration `TeX-AMS-MML_SVG.js`

Thank You,
Alex

**Comment by adam-p** [comment]
I created a new issue for switching to CodeCogs: #261. I intend to do it in the next release.

---

I also had a bad idea: Instead of creating image links to Google Charts or to CodeCogs, etc., create them to an intermediate service (that I run or the user themselves could run). That service would look at the requesting user agent (Gmail in Chrome? Thunderbird? Postbox? Evernote in Firefox?) and decide which rendering to provide (indicating format with mimetype). That way we wouldn't be restricted to the least common denominator -- instead we could return the optimal choice for each platform. (Would have to be an image, though. Not fancy MathJax JS stuff.)

There are a bunch of really good reasons to *not* do this. 1) It adds another privacy compromise. 2) Running a server, bleh. 3) Figuring out and maintaining "optimal" choices.

Can we even tell what is requesting?

Thunderbird receiving:
```http
Accept-Language: en-US,en;q=0.5
Connect-Time: 1
Accept-Encoding: gzip, deflate
Accept: image/png,image/*;q=0.8,*/*;q=0.5
Connection: close
X-Request-Id: e2462045-b85c-4373-bd89-b862dba58a81
Via: 1.1 vegur
Total-Route-Time: 0
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64; rv:31.0) Gecko/20100101 Thunderbird/31.5.0
Host: requestb.in
```

Gmail+Chrome composing:
```http
Accept-Language: en-US,en;q=0.8,ja;q=0.6,ko;q=0.4
Connect-Time: 3
Accept-Encoding: gzip, deflate, sdch
Accept: image/webp,*/*;q=0.8
Host: requestb.in
Dnt: 1
X-Request-Id: 45accc18-7244-406d-8905-9700f8914f3a
Via: 1.1 vegur
Total-Route-Time: 0
User-Agent: Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.89 Safari/537.36
Connection: close
```

Gmail+Chrome receiving:
```http
Connect-Time: 3
Accept-Encoding: gzip,deflate
Host: requestb.in
X-Request-Id: c4259f01-10be-4686-87da-3078cc96ed1a
Via: 1.1 vegur
Total-Route-Time: 2
User-Agent: Mozilla/5.0 (Windows NT 5.1; rv:11.0) Gecko Firefox/11.0 (via ggpht.com GoogleImageProxy)
Connection: close
```

Firefox+Yahoo receiving:
```http
Accept-Language: en-US,en;q=0.5
Total-Route-Time: 0
Accept-Encoding: gzip, deflate
Accept: image/png,image/*;q=0.8,*/*;q=0.5
Connection: close
Dnt: 1
X-Request-Id: c393f6b0-5668-464b-997c-3a23ae16c175
Via: 1.1 vegur
Connect-Time: 0
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64; rv:38.0) Gecko/20100101 Firefox/38.0
Cookie: session=eyJyZWNlbnQiOlsiMWkxb3BkdzEiXX0.B-cuwQ.CQ511AR4VinZPLawrbKAJR_ig24
Host: requestb.in
```

Chrome+Outlook.com receiving:
```http
Total-Route-Time: 0
Accept-Language: en-ca,en,en
Host: requestb.in
X-Request-Id: f8c9a26b-b823-4dd3-92da-8ebaf85a2287
Via: 1.1 vegur
Connect-Time: 2
User-Agent: Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.89 Safari/537.36
Connection: close
```

Chrome+Evernote composing:
```http
Accept: image/webp,*/*;q=0.8
Total-Route-Time: 0
User-Agent: Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.89 Safari/537.36
X-Request-Id: 3a95184f-68f6-47bf-9dcc-533a0bd2e446
Accept-Encoding: gzip, deflate, sdch
Connection: close
Via: 1.1 vegur
Dnt: 1
Host: requestb.in
Accept-Language: en-US,en;q=0.8,ja;q=0.6,ko;q=0.4
Connect-Time: 3
```

Well, that's not hopeful. Thunderbird is obvious from the user agent, Gmail-receiving (but not browser) is obvious from the `GoogleImageProxy`, the browser being used is obvious for the others, but... that's about it. 

Maybe there's some kind of overall fingerprint for each combination? Like, maybe Yahoo+Firefox has a set of headers and header values that can consistently distinguish it from the other services? Seems like a big, fragile hassle.

I think I can safely put this idea out of my head.

**Comment by cben** [comment]
Yeah, sounds pretty painful.

What would be useful to vary in the response?

   -

   Format? Replacing PNG with SVG would be nice, and could be done sanely
   if only clients mentioned SVG in the Accept: header.
   Alas, they don’t, except for — I’m embarassed to say — IE:
   https://bugzilla.mozilla.org/show_bug.cgi?id=824623#c9
    -

   Better resolution for high DPI (“retina”) clients? Nice but math is
   relatively small and grey&white, sending the big image to everyone is
   probably way simpler. IIUC the imgset / picture efforts, header negotiation
   was never considered an option.

User-agent sniffing is I think insane unless you’re full-time maintainer of
a service dedicated to images, e.g.
http://blog.imgix.com/post/90838796454/webp-jpeg-xr-progressive-jpg-support-w-auto
…
​

**Comment by adam-p** [comment]
Mostly for SVG vs. PNG. Yeah, not much gain for quite a bit of pain.

Regarding retina clients: When replacing with CodeCogs, I plan on getting a higher DPI image and then explicitly setting the size on it.

**Comment by adam-p** [comment]
> Regarding retina clients: When replacing with CodeCogs, I plan on getting a higher DPI image and then explicitly setting the size on it.

I revoke that statement. I can't figure out a way to satisfactorily size the higher DPI image (in a way that works everywhere). At least until/unless we have both inline and block math -- then the inline can be `height: 1em`, and the block can be... something else.

**Comment by cben** [comment]
Let's have inline&block then :-)  Anything I can help?

Does `height: 1em` work reliably?  And what about `height: 1ex`?
Inline math is the hard one, as getting x-height/ascending/descending right matters a lot to readability, and I'd love to understand that irrespective of retina.

**Comment by cben** [comment]
Ah, perhaps I see what you mean: math should have different size depending on formula, e.g. `$x$` vs `$x^{a^b}$`.  Charts API and CodeCogs retuns appropriately sized image (in pixels), which automatically looks approximately good size next to specific font (say 16px)...
But setting `height` forces all images to same height, which is wrong, except for very simple math that doesn't ascend/descend.  Did you mean that by "inline"?

How about interrogating the returned image size and scaling the em/ex height appropriately?

**Comment by cben** [comment]
Also, re-reading this thread notice MathML works not only in ThunderBird but also on iOS and Apple Mail!
http://www.maths-informatique-jeux.com/blog/frederic/?post/2012/11/14/Writing-mathematics-in-emails
![mathml ipod](http://www.maths-informatique-jeux.com/blog/all-blogs/public/mathml-in-emails/mathml-ipod.png)

Perhaps MathML with PNG fallback is doable and sufficient.
(I actually don't have any Apple devices, my HiDPI are linuxes and androids, but I applaud Apple for supporting all the good stuff.  And sematically including MathML in the mail is the Right Thing.  And CodeCogs can even [generate it for us](http://www.codecogs.com/latex/latextomathml.php?\sum_0^\infty%20\frac{1}{n}).  Now just to find a fallback method that works in email...)

**Comment by cben** [comment]
> How about interrogating the returned image size and scaling the em/ex
height appropriately?

Configuring this actually works (only tested in Gmail in Chrome though):

<img src="https://latex.codecogs.com/png.latex?%5Cdpi%7B200%7D%20%5Cbg_green%20{urlmathcode}"
alt="{mathcode}" onload="this.style.height=(this.height/12+'ex')">

(testing .height instead of .naturalHeight because it’s more portable and
should be equivalent in a fresh unstyled image [
http://www.jacklmoore.com/notes/naturalwidth-and-naturalheight-in-ie/]
It also works with svg.latex, though needs a different factor.)

It turns:

# header $a$ bar $a^b_c$
### subsubheader $a$ bar $a^b_c$
foo $a$ baz $a^b$ bas $a_c$ bar $a^b_c$ quux

into (I hope GH Issues will let the images through):
header [image: a] bar [image: a^b_c] subsubheader [image: a] bar [image:
a^b_c]

foo [image: a] baz [image: a^b] bas [image: a_c] bar [image: a^b_c] quux

I think the header-relative sizing is already a win!

Moreover, I believe it *should* match any font used by any viewer (that
supports ex height units).
The arbitrary this.height/12 factor (which should perhaps be 13–14)
pertains to the font size produced by CodeCogs.

Removing the debugging \bg_green triggers a strange size anomaly:
header [image: a] bar [image: a^b_c]subsubheader [image: a] bar [image:
a^b_c]

foo [image: a] baz [image: a^b] bas [image: a_c] bar [image: a^b_c] quux

It took me a lot of time to understand why CodeCogs oversizes some small
images.
Here are all the letters:

$a$ $b$ $c$ $d$ $e$ $f$ $g$ $h$ $i$ $j$ $k$ $l$ $m$ $n$ $o$ $p$ $q$ $r$ $s$
$t$ $u$ $v$ $w$ $x$ $y$ $z$

without dpi and without setting style.height set at all:

[image: a] [image: b] [image: c] [image: d] [image: e] [image: f] [image: g]
 [image: h] [image: i] [image: j] [image: k] [image: l] [image: m] [image:
n] [image: o] [image: p] [image: q] [image: r] [image: s] [image: t] [image:
u] [image: v] [image: w] [image: x] [image: y] [image: z]

and with \dpi{200}:

[image: a] [image: b] [image: c] [image: d] [image: e] [image: f] [image: g]
 [image: h] [image: i] [image: j] [image: k] [image: l] [image: m] [image:
n] [image: o] [image: p] [image: q] [image: r] [image: s] [image: t] [image:
u] [image: v] [image: w] [image: x] [image: y] [image: z]

Aha!  CodeCogs really crops to the minimal box containing any paint, which
depends even on the letter used.
That's why using a non-transparent bg helps?

And then what?  They upscale it if it's too small?!  But not for all
letters (why is a big and c small)?

/home/beni/Downloads/a.png: PNG image data, 26 x 26, 4-bit colormap,
non-interlaced

/home/beni/Downloads/b.png: PNG image data, 13 x 24, 4-bit colormap,
non-interlaced

/home/beni/Downloads/c.png: PNG image data, 13 x 15, 4-bit colormap,
non-interlaced
There is more weirdness and non-reproducibility going on, and I suspect
I've done too many thousand requests over the last hour, so I'll stop here
for now.

**Comment by cben** [comment]
Sorry, I hoped gmail rich text could survive into github, but it seems github took the text/plain version
Above comment with unmangled IMG tags: 
http://codepen.io/cben/full/ogQqLP

**Comment by cben** [comment]
Playing with other options in the CodeCogs editor, I think it's a bug on their end, and a bewilderingly random one.  The "inline" checkbox seems to help but some math is still oversized.  Using gif instead of png seemed to help but then I discovered that while `a` and `aq` are normal, `aa` is supersized — even with bg_white:
![gif a](http://latex.codecogs.com/gif.latex?%5Cdpi%7B300%7D%20%5Cbg_white%20a) ![gif aa](http://latex.codecogs.com/gif.latex?%5Cdpi%7B300%7D%20%5Cbg_white%20aa)
(Github sanitizes style attributes here, but I don't need those - the issue is with the images themselves - some have more pixels.)
Another "workaround" that I noticed worked at some point is wrapping {} around the math.  For some reason it mostly helps - but `${n}$` is still supersized...

The bottom line is that standard dpi *seems* to work fine but at this point I'm not sure what to trust.

**Comment by adam-p** [comment]
Epic investigation. I love when someone *else* goes down the rabbit hole.

Your inline JS resize idea is great. A couple things, though: 

1. It doesn't work in the MDH options Preview because of iframe security restrictions. This isn't a problem in Gmail, since it doesn't put the compose area in an iframe (anymore -- it once did), but there are probably other places that use a src-less iframe. Anyway, we could get around it with JS in the extension.

2. For some reason Gmail strips the `height` style and replaces it with `min-height`. However, it leaves and respects `max-height`. So your JS needs to be something like: `this.style.height=this.style.maxHeight=this.style.minHeight=(this.height/18+'ex')`.

The size variation is utterly bizarre. Any chance you want to post to the [CodeCogs forum](http://www.codecogs.com/forum.php), suggesting that it's broken?

In exchange for that effort, I'll offer you this data point and possible solution to our woes: If you render to GIF instead of PNG (so, `gif.latex`), *and* you use `\inline`, the letters look... perfect.

![image](https://cloud.githubusercontent.com/assets/425687/6701382/548bb420-ccf2-11e4-8734-e743a02b370f.png)

But if you don't use `\inline`... rubbish:

![image](https://cloud.githubusercontent.com/assets/425687/6701376/3e827920-ccf2-11e4-922d-79b74297064a.png)

Note that I have a plain letter `x` around the $x$. And I'm using `vertical-align: middle`, which I think does a pretty good (or acceptable) job of making the different heights and below-the-line letters look more sensible.

Here's some more numeric evidence of wackiness (and a cool CodeCogs trick):

```
# 'a' inline at 300dpi
$ curl "http://latex.codecogs.com/gif.json?%5Cinline%20%5Cdpi%7B300%7D%20a"
>> ParseEqn({ "latex": { "type":"gif", "equation":"a", "site":"file", "file":"56129c80c955fda42022169d43317b83_16.gif", "url":"http://latex.codecogs.com/eq/56/12/56129c80c955fda42022169d43317b83_16.gif", "width":"22", "height":"20", "baseline":"1"} });

# 'a' not inline at 300dpi
$ curl "http://latex.codecogs.com/gif.json?%5Cdpi%7B300%7D%20a"
>> ParseEqn({ "latex": { "type":"gif", "equation":"a", "site":"file", "file":"ad9fca8dc07e6221bd68105cee111b86_16.gif", "url":"http://latex.codecogs.com/eq/ad/9f/ad9fca8dc07e6221bd68105cee111b86_16.gif", "width":"23", "height":"22" } });

# 'c' inline at 300dpi
$ curl "http://latex.codecogs.com/gif.json?%5Cinline%20%5Cdpi%7B300%7D%20c"
>> ParseEqn({ "latex": { "type":"gif", "equation":"c", "site":"file", "file":"89173b6c921f925664199cb083b525c5_36.gif", "url":"http://latex.codecogs.com/eq/89/17/89173b6c921f925664199cb083b525c5_36.gif", "width":"19", "height":"20", "baseline":"1"} });

# 'c' not inline at 300dpi
$ curl "http://latex.codecogs.com/gif.json?%5Cdpi%7B300%7D%20c"
>> ParseEqn({ "latex": { "type":"gif", "equation":"c", "site":"file", "file":"4a63d47c580007a4b45bd7b4325a7f88_36.gif", "url":"http://latex.codecogs.com/eq/4a/63/4a63d47c580007a4b45bd7b4325a7f88_36.gif", "width":"40", "height":"47" } });
```

'a' is 20 vs. 22. 'c' is 20 vs. 47. What the hell.

(I learned about the `gif.json` thing in [this CodeCogs forum post](http://www.codecogs.com/pages/forums/pagegen.php?id=2737). Also note the `"baseline"` in the `inline` response. It's `1` for 'c' and `4` for 'j', for example. It's also `4` for the quadratic equation. I think that, in theory, we could use that value for shifting the images. However, I'm not sure how to use that value. Setting a negative `margin-bottom` gets stripped by Gmail. And... I don't know any other ways.)

---

Recap:

Inline GIF ftw.

You should totally post to the CodeCogs forum, @cben.

**Comment by adam-p** [comment]
More `vertical-align` stuff...

I modified the test alphabet by adding the plain letter beside each math letter.

With no `vertical-align` set (so using the `baseline` default):
![image](https://cloud.githubusercontent.com/assets/425687/6701629/e34fda26-ccf5-11e4-8928-6bec6b716d4d.png)

With `vertical-align: middle`:
![image](https://cloud.githubusercontent.com/assets/425687/6701652/27891162-ccf6-11e4-8ed4-4111a03133ee.png)

(I fooled around with some other `vertical-align` values, but they look worse.)

Tall letters -- like 't' and 'k' -- get worse, but the letters with "[descenders](http://english.stackexchange.com/questions/18794/is-there-a-name-for-letters-that-cross-the-line-hang-low-when-written-in-low)" -- like 'j' and 'y' -- look a lot better. 

I think it's a net gain...? But barely.

**Comment by adam-p** [comment]
Regarding what I meant by "inline" and "block": I mean the former in the sense of "inline with normal text" and the latter to mean "standing by itself". (Say we had say `$...$` for inline math and `$$...$$` for block math.)

I would be pretty comfortable forcing the inline math to `1em` (kind of thing) and then letting the block math be whatever height makes sense (`height*300/110`, say, if we use 300dpi and 110dpi is "normal").

(In theory, I suppose, the code could detect math that is by itself in a paragraph versus math that's in a paragraph with other text. Hmm. Seems like a lot of work, though.)

(For the record, I never use the math feature, so it's all hypothetical to me. People who actually use it have more valuable opinions.)

P.S. This is really the wrong issue for these discussions. But too late now...

**Comment by adam-p** [comment]
Receiving an email in Thunderbird with the height set (either just `height` or `max-height` etc.) looks crappy:

![image](https://cloud.githubusercontent.com/assets/425687/6701991/93922b4c-ccfa-11e4-8af1-bfe67eb7de6b.png)

Without the height set is not similarly cut off:

![image](https://cloud.githubusercontent.com/assets/425687/6701995/a2e97c1c-ccfa-11e4-9a0b-3c95dfc7b53c.png)

This high-dpi stuff is making me insane.

**Comment by cben** [comment]
> P.S. This is really the wrong issue for these discussions. But too late now...

Oops.  Will continue on #261.  Will report on CodeCogs forum.

**Comment by upsidedowntext** [comment]
[backwards text generator ](https://www.abtextgenerator.com/backwards-text/)

**Comment by drdrjacobs** [comment]
So are we any closer these days to the HTML + CSS promised land?

**Comment by adam-p** [comment]
I'm going to close this issue. There's great info here that I'm sure we'll refer to in the future, but it's too sprawling.

The landscape of TeX rendering (and what is supported in email) hasn't really changed much since we started this conversation, unfortunately. On the other hand, there's a glimmer of hope for client-side rendering. Follow #874 for that.
